File: Data Structure/Array or Matrix/Medium/LET_no54_Spiral Matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        res = []
        i, j, di, dj = 0, 0, 0, 1
        m, n = len(matrix), len(matrix[0])
        
        for _ in range(m * n):
            res.append(matrix[i][j])
            matrix[i][j] = -1000
            if matrix[(i + di) % m][(j + dj) % n] == -1000:
                di, dj = dj, -di
            i, j = i + di, j + dj
        
        return res

File: Data Structure/Array or Matrix/Medium/LET_no54_Spiral Matrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        int m = matrix.size(), n = matrix[0].size();
        int i = 0, j = 0, di = 0, dj = 1, times = m * n;
        vector<int> res;
        
        while (times--) {
            res.push_back(matrix[i][j]);
            matrix[i][j] = -1000;
            if (matrix[(m + i + di) % m][(n + j + dj) % n] == -1000) {
                int temp = dj;
                dj = -di;
                di = temp;
            }
            i += di; j += dj;
        }
        
        return res;
    }
};